Hypergrid 3.0 prep

joneit opened the pull request:
Implementation notes follow.

> **See:** Issue #614: _Proposal for slimmed down data model_

> **Note:** ~I'm just trying this technique out to see how it goes: This branch is currently based on the unmerged #648 branch but to make reviewing easier I temporarily based the PR on #648 as well, which successfully hides it's edits from the diff. ¶ We do not of course want to merge it to #648's branch; I'll set the base back to `develop` once it's merged.~ _¶ This note is now outdated; #648 has now been merged; I have now reset this PR's base back to `develop` — however there are now merge commits I need to resolve (due to other merges to `develop`). All in all this technique was not worth it!_

## Data source externalized

In addition to codifying the data source interface, the data source module is now fully externalized with a black-box interface with minimal requirements. Hypergrid makes no assumptions about the data source's design or implementation details, just so long as it conforms to the minimum interface.

The data source can now be specified as an option to the DataModel constructor (which can come from the behavior constructor (which can come from the Hypergrid constructor)).

See [_Data source interface spec (proposed)_](https://github.com/fin-hypergrid/core/wiki/Data-source-interface-spec-(proposed)) (a wiki) for details.

## Row and cell properties

Row and cell properties are row-specific things that necessarily parallel the row data (as "row metadata" or simply "metadata"). The metadata can be stored in the data rows themselves or in a separate parallel data structure. Storing it with the data rows makes it especially convenient to persist the metadata along with the actual row data.

See the [_Row and cell properties as metadata_](https://github.com/fin-hypergrid/core/wiki/Row-and-cell-properties-as-metadata) wiki for details.

## Module Loader

The Hypergrid Module Loader implementation (`Hypergrid.require()`) has been expanded somewhat.

See the [_Client Modules_](https://github.com/fin-hypergrid/core/wiki/Client-Modules) wiki for details.

To encourage the use of this built-in module loader, direct access to selected internal modules has been deprecated with a warning that it will be removed in v4.

> NOTES:
> 1. All of the `require`s below will work as is on the client side from within a Hypergrid Client Module or from within an unwrapped script by using `fin.Hypergrid.require` instead.
> 2. `Hypergrid.require` does not expect `require` strings to end in `/`, `.js`, `/index`, or `/index.js` so please omit.

### Accessing internal modules

Deprecated usage<br>_(issues warning)_ | Favored usage
---|---
`fin.Hypergrid.lib.module` | `require('fin-hypergrid/src/lib/module')`
`fin.Hypergrid.features.module` | `require('fin-hypergrid/src/features/module')`
`fin.Hypergrid.behaviors.module` | `require('fin-hypergrid/src/behaviors/module')`
`fin.Hypergrid.dataModels.module` | `require('fin-hypergrid/src/dataModels/module')`
`fin.Hypergrid.images` | `require('fin-hypergrid/images')`
`fin.Hypergrid.Base` | `require('fin-hypergrid/src/Base')`

For completeness `require('fin-hypergrid/src/defaults')` also works although `Hypergrid.defaults` (synonym: `Hypergrid.properties`) remains the favored usage.

### Accessing external modules

The following modules are all part of the build and can be accessed without loading them separately with `<script>` tags:

Deprecated usage | Favored usage
---|---
`window.datasaur.base`<br>_(throws error)_ | `require('datasaur-base')`
`fin.Hypergrid.lib.DataSourceOrigin`<br>_(throws error)_ | `require('datasaur-local')`
`fin.Hypergrid.lib.fields`<br>_(issues warning)_ | `require('fin-hypergrid-field-tools')`
`fin.Hypergrid.rectangular`<br>_(issues warning)_ | `require('rectangular')` | Deprecated usage issues warning.
_(Not previously accessible)_ | `require('extend-me')`
_(Not previously accessible)_ | `require('scrollbar')`<br>_(a.k.a `finbars`)_
_(Not previously accessible)_ | `require('mustache')`
_(Not previously accessible)_ | `require('object-iterators')`
_(Not previously accessible)_ | `require('overrider')`

> **Special note regarding the two "datasaur" modules.** _(This note pertains only to applications using the build file from the CDN.)_ Although the two "datasaur" modules are still in the build _and_ the local data source is still used by default, this usage is deprecated as of 3.0.0 and a console warning will be issued. In a future version, they will be removed from the build and there will be no default data source. The fact that it still included in the 3.0.0 build and still used by default is purely a courtesy to help wean developers off of depending upon it. Developers should start getting used to this now by "bringing their own" data source. We have made this easy by wrapping them as Hypergrid Client Modules on the CDN. What this means is that the `require` (or `Hypergrid.require` as the case may be) syntax will continue to work, but the modules will need to be loaded explicitly with `<script>` tags.


## New property: `Base.prototype.version`

A new property `version`, containing the Hypergrid module version number from package.json, is now available to all instances of classes extended from `Base`.


## New method: `Behavior.prototype.addState`

Similar to `setState` except does not clear the state (_i.e.,_ grid properties) first.

Updated to use `Object.assign()` rather than a polyfill loop.

Also, moved `createColumns()` call from `setState` to `clearState` because columns carry state (_i.e.,_ column properties).


## Missing return values

Added the missing (though non-essential) `return`s to `setCellProperties` and `addCellProperties `.


## Themes

Moved all code dealing with themes from Hypergrid.js to a new file themes.js.

### New method: `registerThemeCollection()`
Registers a whole hash of themes all at once. Just a one liner with a loop.

### New repo: [`fin-hypergrid-themes`](https://github.com/fin-hypergrid/themes)
Also created a new repo and moved the unused buildTheme.js file to it. This file contains the `buildTheme` function which is verbatim from the original polymer version of Hypergrid (except renamed from `buildPolymerTheme`). It builds a Hypergrid theme object from a loaded Polymer Theme CSS stylesheet. It is still useful since you can utilize a Polymer Theme without having to be a Polymerized app.

Also included in the repo are:
* The `theme` folder from the Polymer version containing the 16 Polymer Theme CSS stylesheet files.
* A folder of 16 Hypergrid theme objects pre-built from the stylesheets.

To build a new theme object from a loaded and enabled Polymer Theme CSS stylesheet:
```js
var Hypergrid = require('fin-hypergrid');
var buildTheme = require('fin-hypergrid-themes').buildTheme;
Hypergrid.registerTheme(buildTheme());
```
To register all the pre-built theme objects:
```js
Hypergrid.registerThemeCollection(require('fin-hypergrid-themes'));
```
And to consume a registered theme:
```js
Hypergrid.applyTheme('candy'); // apply the Candy theme as a "global theme" for grid instances
myGrid.applyTheme('dark-side'); // apply the Dark-Side them as a "local theme" overlaid on the global theme
```
See the [README](https://github.com/fin-hypergrid/themes/blob/master/README.md) for details and some more examples.


## Miscellaney

* `package.json`
   * `chai` was erroneously being included as a dependency (!); it is now a dev-dependency. (No biggie since there were no references to it.)
   * Updated `mustache` to version 2.3 to avoid reported vulnerability (#638 ).
* Moved and renamed Hypergrid.js
The main file, previously `./src/Hypergrid.js` is now in its own folder and has been renamed to `./src/Hypergrid/index.js`. Several files required only by this one have also been moved into that folder (`events.js`, `selections.js`, `themes.js`).
* Renamed build file entry point from `index.js` to `build.js` as this is more descriptive and it is never actually `require`d anywhere.
* Wrapped mix-in exports in a `mixin` property.
* Added `require('finbars')` to `Hypergrid.modules.scrollbar` which is how Hypergrid now references it internally. This makes `finbars` overridable while at the same time accessible as an external module — albeit with a pseudonym — as `require('scrollbar').

joneit commented:
@Dwaynekj Let's coordinate, please, going forward. Is there some objection on your end to this PR?

joneit commented:
@dcchuck As you know I'll be contributing from my new position at JPMC. I believe we are still in agreement regarding v3.0, no? Also, my write access has evaporated!

Dwaynekj commented:
No objection, I wasn't aware of your new role. I just thought this PR has been left lingering

Dwaynekj commented:
As for the role change, I'll explain over text

joneit commented:
@Dwaynekj Thanks for revivifying :smile:.

I linked to the issue in the description above. (I can't figure out how to re-open the issue, which you closed last night. Do you know how?)

Shall we remove the "WIP" and throw it open to the community for review.

Dwaynekj commented:
@joneit Reopened the issue. LMK if your still having trouble
Yea if your ready for a design review, let's do that

joneit commented:
@nmichaud @Dwaynekj @dcchuck @neilslinger Ready for review. Do leave feedback in the Files Changed tab. I realize the PR is hefty so I'm prepared to do a walk-through at our upcoming meet. I'll also update the PR description above with a brief outline of features. _Jonathan_

Dwaynekj commented:
@joneit This is an big/expected change. How do you want us to evaluate?

joneit commented:
@Dwaynekj Wait for my implementation notes.

Dwaynekj commented:
@joneit Sure this week? It seems your adding more changes

joneit commented:
@Dwaynekj @dcchuck @nmichaud @neilslinger @texodus @stevewirts Hopefully everyone can find the time to read through the implementation notes above, _including the issues/wikis referenced therein._

In the _Files Changed_ tab, please leave comments and concerns by clicking next to line numbers. Please use the **Start Review** button (rather than the Leave Comment button) and submit your review when done with the **Review Change** button at top of page followed by the revealed **Submit Review** button.

I know it looks huge, but much of the diff consists of tweaks to comments, a good amount of removed code (yea!), and moved code. I would be more than happy to walk you through the substantive changes hidden amongst the noise.

Thanks, _JE_

PS: Our meet on Wednesday is mostly about discussing planned development activity for the immediate future (beyond the current PRs). [Iff](https://en.wikipedia.org/wiki/IFF) time permits I could attempt a summary walk-through then.

joneit commented:
Rebased after merging #643.

joneit commented:
Per team mtg of 12/20/17, I'll be adding another commit to:
* Make data source `setData` completely optional, to be called only when `data` option provided to constructor.
* Remove datasaur modules from the build. There will be no default datasource; developer must now provide. This will make quickstart slightly bulkier but was judged to be worth it.
* We still need to figure out status of "special column" support in core.

nmichaud commented:
Why was this closed?

On Jan 10, 2018 8:27 AM, "Jonathan Eiten" <notifications@github.com> wrote:

> Closed #619 <https://github.com/fin-hypergrid/core/pull/619>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fin-hypergrid/core/pull/619#event-1417661815>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAnj_l9_jVN-ItTrjGx_2GZzqs1VHCY0ks5tJLq5gaJpZM4O5ZJm>
> .
>

joneit commented:
Reopening; closed in error.

@nmichaud Thanks for the heads up. It was actually closed automatically by GitHub when I deleted the branch on which it was based.

(I had based it on a "tmp" branch that included #648 and #661 so that this PR's diff would not show those edits so as not to confuse reveiwers. Just a technique I was playing with. Obviously has its drawbacks. I have now changed the base branch back to `develop`.)

joneit commented:
I just rebased this to the 2.1.0 release and squashed it. I am now working on implementing the recommendations from the 12/20/2017 meet (see above).

stevewirts commented:
I've updated the iver-tree repo in my GitHub account
I depolymerized things and it's easy to run and see the "cleaner"
view-model separation

On Jan 22, 2018 5:03 PM, "Jonathan Eiten" <notifications@github.com> wrote:

> I just rebased this to the 2.1.0 release and squashed it. I am now working
> on implementing the recommendations from the 12/20/2017 meet (see above).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fin-hypergrid/core/pull/619#issuecomment-359583073>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC6Sp_8JI8TWf68yzr3mXYqNZq28ZgXZks5tNQWqgaJpZM4O5ZJm>
> .
>

nmichaud commented:
Steve - is there an example of iver-tree that doesn't require a q server?

On Mon, Jan 22, 2018 at 5:14 PM, stevewirts <notifications@github.com>
wrote:

> I've updated the iver-tree repo in my GitHub account
> I depolymerized things and it's easy to run and see the "cleaner"
> view-model separation
>
> On Jan 22, 2018 5:03 PM, "Jonathan Eiten" <notifications@github.com>
> wrote:
>
> > I just rebased this to the 2.1.0 release and squashed it. I am now
> working
> > on implementing the recommendations from the 12/20/2017 meet (see above).
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/fin-hypergrid/core/pull/619#issuecomment-359583073>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AC6Sp_
> 8JI8TWf68yzr3mXYqNZq28ZgXZks5tNQWqgaJpZM4O5ZJm>
>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fin-hypergrid/core/pull/619#issuecomment-359587608>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnj_iN9XaIK56UF9qXXsPoeyqP3wY_3ks5tNQgzgaJpZM4O5ZJm>
> .
>



-- 
-----------------------------------
Naveen Michaud-Agrawal

stevewirts commented:
no unfortunately not, the point was to have a simple example of a virtual
server.
It's easy to setup though.

1) install free q;  https://kx.com/download/
2) cd into the iver-tree/q directory, run "q sortable.q"
3) select the simpletable.html page from "npm run dev" (this starts up
webpack dev server and will launch a browser)



On Mon, Jan 22, 2018 at 9:37 PM, Naveen Michaud-Agrawal <
notifications@github.com> wrote:

> Steve - is there an example of iver-tree that doesn't require a q server?
>
> On Mon, Jan 22, 2018 at 5:14 PM, stevewirts <notifications@github.com>
> wrote:
>
> > I've updated the iver-tree repo in my GitHub account
> > I depolymerized things and it's easy to run and see the "cleaner"
> > view-model separation
> >
> > On Jan 22, 2018 5:03 PM, "Jonathan Eiten" <notifications@github.com>
> > wrote:
> >
> > > I just rebased this to the 2.1.0 release and squashed it. I am now
> > working
> > > on implementing the recommendations from the 12/20/2017 meet (see
> above).
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/fin-hypergrid/core/pull/619#issuecomment-359583073
> >,
> > > or mute the thread
> > > <https://github.com/notifications/unsubscribe-auth/AC6Sp_
> > 8JI8TWf68yzr3mXYqNZq28ZgXZks5tNQWqgaJpZM4O5ZJm>
> >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/fin-hypergrid/core/pull/619#issuecomment-359587608>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AAnj_
> iN9XaIK56UF9qXXsPoeyqP3wY_3ks5tNQgzgaJpZM4O5ZJm>
> > .
> >
>
>
>
> --
> -----------------------------------
> Naveen Michaud-Agrawal
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fin-hypergrid/core/pull/619#issuecomment-359654636>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC6Sp_hepyTvm0dvss8dFetrD4abcUx7ks5tNUXygaJpZM4O5ZJm>
> .
>

joneit commented:
Removed 481 lines (22 KB). Woo-hoo!
```bash
git diff HEAD^ HEAD | egrep '^-' | wc
     481    2668   22819
```

joneit commented:
Re: edac961

Only events on data model can be triggered by data source.

The only such event currently defined is `data-changed` — which calls the new grid method `fireDataChangedEvent` which:
1. calls `grid.repaint()`
2. fires the new grid event `fin-data-changed`.

joneit commented:
Re: f132899

You can now pass a fully instantiated data source to `Hypergrid` constructor (or `grid.behavior.reset` or `grid.behavior.dataModel.reset`) in options object as `options.dataSource`.

If given, `options.data` and `options.schema` are ignored.

Iff it differs from the current data source (`grid.behavior.dataModel.dataSource`), the row accessor (returned by the `getRow()` fallback) is recreated.

joneit commented:
Rebased squashed to `2.1.2`.

joneit commented:
Closing this PR in preparation for posing a cleaner v3 PR rebased to latest v2 and much cleaner.
